### PR TITLE
Add ExceptionOps to extract exception details safely

### PR DESCRIPTION
**Pull request by leo-da**
`Throwable.getMessage` can return `null` which caused unexpected NPE
in JSON API error handling

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags, if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/util/ExceptionOps.scala [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+package com.digitalasset.http.util
+
+object ExceptionOps {
+
+  implicit class ExceptionOps(val underlying: Throwable) extends AnyVal {
+    def description: String = getDescription(underlying)
+  }
+
+  def getDescription(e: Throwable): String = {
+    val name: String = getClass.getName
```
Comment (stefanobaghino-da) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/util/ExceptionOps.scala]:
```suggestion
    val name: String = e.getClass.getName
```
Are you sure this is not what you're trying to do? Otherwise I believe name would always be `com.digitalasset.http.util.ExceptionOps$`.

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/util/ExceptionOps.scala [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    Option(e.getMessage).filter(_.nonEmpty) match {
```
Comment (stefanobaghino-da) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/util/ExceptionOps.scala]: Is there a specific reason why you'd always prepend the class name? Why not doing that only if the message is missing? If you need to prepend the class name you can do it at call site.

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/util/ExceptionOps.scala (hunk 1 above)
Comment (leo-da) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/util/ExceptionOps.scala]: @stefanobaghino-da good catch! Thanks!

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/MapKOps.scala [hunk 3]
```diff
@@ -8,7 +8,8 @@ import scala.collection.GenTraversableOnce
 import scala.collection.immutable.{Map, MapLike}
 
 /** Halfway between the *-kinded MapLike and *->*->*-kinded MapOps. */
-trait MapKOps[K, +V, +This[+TV] <: Map[K, TV] with MapKOps[K, TV, This]] extends MapLike[K, V, This[V]] { this: This[V] =>
+trait MapKOps[K, +V, +This[+TV] <: Map[K, TV] with MapKOps[K, TV, This]]
+    extends MapLike[K, V, This[V]] { this: This[V] =>
```
Comment (SamirTalwar) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/MapKOps.scala]: Seems unrelated (and there are a few others too). How come?

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/util/ExceptionOps.scala (hunk 2 above)
Comment (leo-da) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/util/ExceptionOps.scala]: I actually want to provide the context, the exception class name. I could just use `Throwable#toString` but anyone can override id, so I don't really control it. with `getDescription` I control exactly how I want to report exceptions. If exception class name makes too much noise we could change it later, but I believe it will be useful for DABL team, they extend JSON API.

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/MapKOps.scala (hunk 3 above)
Comment (leo-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/MapKOps.scala]: don't know... `./fmt.sh` did this.
